docs/advanced.rst

```diff
@@ -0,0 +1,72 @@
+Advanced Usage
+==============
+
+The workspace adapter
+---------------------
+
+The app provides an adapter that you can use to provide extra, customized data about a workspace.
+
+First, you will need to define a new model with the additional fields as well as a one-to-one field to :class:`~anvil_consortium_manager.models.Workspace`, e.g., ``WorkspaceData``. This one-to-one field should be called ``workspace``.
+
+.. code-block:: python
+
+    from django.db import models
+    from anvil_consortium_manager.models import Workspace
+
+    class CustomWorkspaceData(models.Model):
+        study_name = models.CharField(max_length=255)
+        consent_code = models.CharField(max_length=16)
+        # This field *must* be called "workspace".
+        workspace = models.OneToOneField(Workspace, on_delete=models.CASCADE)
+
+
+You should also define a form containing the additional fields, and excluding the ``workspace`` field.
```

Comment:
You should probably validate the form is set up correctly to exclude the workspace field in the adapter. I see why you are doing it this way, given your custom multi form view implementation. I think it is fine to leave it that way, although you could probably restructure to allow workspace to be a defined form field. Especially as this allows custom workspace cleaning to be established in the custom form if needed.

Reply:
It looks possible to restructure the view. How would I display the forms in the view but also hide the workspace field of the workspace_data_form? I could populate it using the workspace that is created before cleaning the workspace_data_form?

Reply:
Ya, the advantage of using the inlin formset stuff is that it does this for you as it determines and validates the relation and 'does the right thing' with the the related form fields display and form fields data. I saw your note about it not working for the workspace import, I wonder if we could work around that somehow - I will take another look at that code.

Alternately, what you suggest is what I was thinking. Making workspace a hidden field and populating using the workspace that is created although we may want to look at how the inline formsets code does it in order to avoid issues.

Reply:
I can at least implement it via inline formsets for the WorkspaceCreate view, but if you look over WorkspaceImport and figure out a clever say to use inline formsets then I am happy to switch that one too.

Reply:
(mostly as a note to myself for next week) For WorkspaceImport: I wonder if defining a Workspace form with no fields would let us use the inline formset.

Reply:
Ya, some variation on this idea should work. Another option would be to leave the fields there, but make them hidden, then set them from your additional 'import_workspace' field on submit but before validation or client side if you prefer. Then any existing field cleaning would still work.

Reply:
I ended up making the workspace field hidden instead of trying to use formsets. One downside is that it makes an API call before it validates the workspace data form, which is not ideal, since it needs to set the workspace field to the workspace that was just created and/or imported before validating.

Reply:
And now, switched to inline formsets. The same issue with making the API call before validating the workspace data form still exists for `WorkspaceImport` but I don't think there's a way around that.